src/stcal/resample/resample.py

```diff
@@ -0,0 +1,1506 @@
+        log.info(f"Driz parameter weight_type: {self.weight_type}")
+
+        self.check_memory_requirements(allowed_memory)
+
+        log.debug('Output mosaic size: {}'.format(self._output_wcs.pixel_shape))
+
+    @property
+    def output_model(self):
+        return self._output_model
+
+    def process_kwargs(self, kwargs):
+        """ A method called by ``__init__`` to process input kwargs before
+        output WCS is created and before output model template is created.
+
+        Returns
+        -------
+        kwargs : dict
+            Unrecognized/not processed ``kwargs``.
+
+        """
+        return {k : v for k, v in kwargs.items()}
+
+    def _warn_extra_kwargs(self, kwargs):
+        for k in kwargs:
+            log.warning(f"Unrecognized argument '{k}' will be ignored.")
+
+    def check_memory_requirements(self, allowed_memory):
+        """ Called just before '_pre_run_callback()' is called to verify
+        that there is enough memory to hold the output. """
+        if allowed_memory is None and "DMODEL_ALLOWED_MEMORY" not in os.environ:
+            return
+
+        allowed_memory = float(allowed_memory)
+
+        # get the available memory
+        available_memory = psutil.virtual_memory().available + psutil.swap_memory().total
+
+        # determine data type of the output model:
+        out_model = self.new_model((2, 2))
+        data = self.get_model_array(out_model)
+        data_type = data.dtype
+        del data, out_model
+
+        # compute the output array size
+        npix = npix = np.prod(self._output_array_shape)
+        nmodels = len(self._input_models)
+        nconpl = nmodels // 32 + (1 if nmodels % 32 else 0)
+        n_arr = self.n_arrays_per_output + 2  # 2 comes from pixmap
+        required_memory = npix * (n_arr * data_type.itemsize + nconpl * 4)
+
+        # compare used to available
+        used_fraction = required_memory / available_memory
+        if used_fraction > allowed_memory:
+            raise OutputTooLargeError(
+                f'Combined ImageModel size {self._output_wcs.array_shape} '
+                f'requires {bytes2human(required_memory)}. '
+                f'Model cannot be instantiated.'
+            )
+
+    def _compute_output_wcs(self, **wcs_pars):
+        """ returns a diustortion-free WCS object and its pixel scale """
+#        owcs = output_wcs_from_input_wcs(self._input_s_region_list, **wcs_pars)
+        owcs = output_wcs_from_input_wcs([self._input_img1_wcs], **wcs_pars)
+        return owcs['output_wcs'], owcs['pscale']
+
+    def preload_input_meta(self, wcs1, filename, s_region):
+        # set-up lists for WCS and file names
+        self._input_img1_wcs = None
+        self._input_s_region_list = []
+        self._input_filename_list = []
+
+        # loop over only science exposures in the ModelLibrary
+        # sci_indices = self._input_models.ind_asn_type("science")
+        with self._input_models:
```

Comment:
We seem to be in the same infinite loop as before here, about whether or not to make certain datamodels dependencies of stcal, either implicitly or explicitly.  It seems to me that with this and other code lines, ResampleBase can only be used if the input models are assumed to be ModelLibrary instances.  Are we okay with this?  If so, should we just make this explicit in some way by specifying the input model type inside the `ResampleBase` abstract base class, and make stpipe a dependency?

Reply:
> We seem to be in the same infinite loop as before here, about whether or not to make certain datamodels dependencies of stcal, either implicitly or explicitly. It seems to me that with this and other code lines, ResampleBase can only be used if the input models are assumed to be ModelLibrary instances.

True. It is difficult to maximize the common code ported to `stcal` with the purpose of reducing maintenance burden in the pipelines given the rigid usage imposed by the `ModelLibrary` (by contrast, `ModelContainer` was behaving like a standard list). I can try to see how to bury `ModelLibrary` into the "IO" class, but it's not going to be pretty.

> Are we okay with this? If so, should we just make this explicit in some way by specifying the input model type inside the `ResampleBase` abstract base class, and make stpipe a dependency?

I don't know what are the plans for these packages but I do not think they are intended to be general purpose algorithms like those in `astropy` (or even `drizzle`) so I do not see an issue with using structures shared by all pipelines.